### PR TITLE
Expose public API for transformation.

### LINKED ISSUE #13763: Expose visitors as public API
Expose the `visitNode`, `visitNodes`, and `visitEachNode` functions as a public API

### LINKED ISSUE #13764: Expose transformations as a public API
Expose the `transformFiles` and `getTransformers` functions as a public API.

### PR DESCRIPTION
**Pull request by rbuckton**
This PR contains the following changes:

* `TransformationContext` is now public:
  * `getEmitResolver` and `getEmitHost` are still marked `/*@internal*/`.
  * Added assertions to ensure transformation hooks are only modified at the right times during transformation.
* `TransformationResult` is now public.
* `Transformer` is now public.
* `visitNode()`, `visitNodes()`, and `visitEachChild()` are now public.
  * NOTE: `visitEachChild()` does not recursively visit `TypeNode` nodes, as they have so far not been necessary for emit transformations. 
* `visitLexicalEnvironment()`, `visitParameterList()`, and `visitFunctionBody()` helpers are now public.
* Added the ability to attach synthetic leading and trailing comments to a node.
* Added a `customTransformers` parameter to `Program.emit()`, to run custom transformations before and after the main transformation pipeline.
* Added a `getCustomTransformers()` method to `LanguageServiceHost` to allow implementers to supply custom transformations to be run before and after the main transformation pipeline.
* Added a `transform()` function in services that can be used to perform arbitrary transformations on source files.
  * NOTE: As the built-in transformers require the `EmitResolver` and `EmitHost`, they are not run as part of `transform()`, as it does not have access to these. If you require this use case consider using `Program.emit()` with the `customTransformers` parameter, or create a `LanguageService` from a custom `LanguageServiceHost` that implements the `getCustomTransformers()` method.

Fixes #13764
Fixes #13763

### CODE REVIEW

File: src/compiler/comments.ts [hunk 1]
```diff
@@ -117,12 +106,89 @@ namespace ts {
...
+
+            emitNodeWithNestedComments(hint, node, emitFlags, emitCallback);
+
+            const trailingComments = emitNode && emitNode.trailingComments;
+            if (some(trailingComments)) {
+                if (extendedDiagnostics) {
+                    performance.mark("postEmitNodeWithSynthesizedComments");
+                }
+
+                debugger;
```
Comment (mhegazy) [inline src/compiler/comments.ts]: remove.

**Comment by mhegazy** [comment]
//CC @alexeagle, @chuckjaz, @mprobst, @rkirov and @evmar would like to get your feedback on this change. this should allow tools like https://github.com/angular/tsickle to inject a transom either before or after the other TS transforms are applied.

//CC @ivogabe, this should supersede https://github.com/Microsoft/TypeScript/pull/11561

**Comment by evmar** [comment]
Thanks for thinking of me!  I intend to look at this but I have had a lot of other work recently.  If you can't wait I also don't think it'd be too harmful to land this and let us poke at it in a release.  I'm not sure you have a mechanism for clearly tagging "experimental" API though.

**Comment by rkirov** [comment]
Unordered observations from my interaction with the API. Apologies if some stem from my misuse or misunderstanding of the API:

- API is subtle, we can use some minimal documentation. I am confused about `hints`, `partiallyEmittedExpression`, and ran into errors of the sort `Lexical environment is suspended` when playing with the `visit*` utilities.
- Copying the tests included, I did a mini-tsickle using purely `addSynteticLeadingComment`. The real implementation would not be so simple, because we always have to check and augment an existing jsdoc. That is the transformer will almost always have to modify the AST, and not just augment.
- Completely failed to write the tsickle pass that rewrites 'export *' (needed because closures module system does not suport that) - https://github.com/angular/tsickle/blob/master/src/tsickle.ts#L402 . Despite `context.enableSubstitution(ts.SyntaxKind.ExportDeclaration)`, I do not see `onSubstituteNode` being called on any such nodes.

**Comment by rbuckton** [comment]
@rkirov While I still have some work to do on the API documentation, here are a few pointers based on your bullet points:

* `hint` - TypeScript uses the same `Identifier` node for identifiers that are part of an expression and identifiers that are part of a declaration. When performing substitution, it is often important to know which context you are in. You should never modify the `hint` that is passed in to the `onEmitNode` or `onSubstituteNode` hooks, but you can inspect it to understand the context you are in:
  * `EmitHint.SourceFile` - Only used for a `SourceFile` node, this hint is used by the emitter to control how comments and source maps are emitted at the top and bottom of a source file. Generally you won't need to use it.
  * `EmitHint.IdentifierName` - Only used for an `Identifier` that is the name of a function or class declaration, this hint is used by the emitter to control how source maps are emitted for the name.
  * `EmitHint.Expression` - Used when emitting an expression, primarily helps to distinguish an `Identifier` that is in an expression position vs. one used in a declaration.
  * `EmitHint.Unspecified` - Used for every other node.
* `PartiallyEmittedExpression` - This is a synthetic node added by transformations to apply comments and source maps for nodes that are partially removed during a transformation. For example, when TypeScript elides a `TypeAssertionExpression`, we replace it with a `PartiallyEmittedExpression` to preserve comments and source maps for the original expression.
* `visitParameterList`/`visitFunctionBody` - When visiting a function-like declaration, we introduce a new lexical environment for the purposes of hoisting generated variable or function declarations. Due to how lexical environments and scoping rules work in JavaScript, the new lexical environment is introduced when we visit the parameter list of a function-like declaration, and is exited after the end of the function body. To ensure lexical environments are nested correctly, we can suspend any modification of a lexical environment until some point at which it is resumed, allowing us to ensure the correct state is maintained when updating the tree. It is not required to use these helpers when visiting a function, and despite their general utility I may consider removing them from the public surface area to avoid confusion. Generally you want to use `visitNode`, `visitNodes`, and `visitEachChild`.
* `onSubstituteNode` - It depends on how you are using the transforms API. This hook provides just-in-time substitution during the final print phase, and is primarily used to substitute a relatively small subset of nodes to avoid an expensive full walk of the tree during every transformation. Since this runs during the print phase, it is quite possible some other transformation may have already replaced the `ExportDeclaration` with some other statement.

**Comment by rkirov** [comment]
Thanks @rbuckton, I created mini implementations of all the transformations we want to do in tsickle and they seem to work great.

The final two questions:
- how can one remove leading comments (tsickle will need to remove existing ones, before adding the new one)
- how would transformers interact with source maps generation?

and one observation, TS internal API are still not strictNullChecks compatible, I had to use a lot of `undefined as any` when using `ts.create*` :)

**Comment by rbuckton** [comment]
@rkirov:
> * how can one remove leading comments (tsickle will need to remove existing ones, before adding the new one)

There is an exported function `setEmitFlags` that you can use for this purpose. It only affects comments read from the source file, not synthesized comments:

```ts
ts.setEmitFlags(node, EmitFlags.NoLeadingComments);
```

> * how would transformers interact with source maps generation?

There is an exported function `setSourceMapRange` that you can use for this:

```ts
// range is a TextRange, either an existing node or some other `{ pos, end }` value.
ts.setSourceMapRange(node, range);
```

**Comment by Busyrev** [comment]
Great feature. Is there any axamples, how to simply load typescript code, transform and emit typescript code to new ts file? I'm playing aroud typescript to typescript code transformations.

**Comment by mhegazy** [comment]
@Busyrev we are working on some documentation and samples. so stay tuned.

**Comment by fabiandev** [comment]
Thanks for the great work!
Did anyone already play around with this and can share a simple example as requested by @Busyrev?
And is there a chance we can get an estimation on when the documentation will be published, or to have a look at a draft version, @mhegazy?

**Comment by alexeagle** [comment]
I made a minimal example for transforms and plugins (in my case I use them
from typescript.js, as part of a wrapper around tsc, rather than in
tsserver.js for use in an editor):
https://github.com/alexeagle/ts_plugin_prototype

**Comment by dsherret** [comment]
@fabiandev here's [another example](https://github.com/dsherret/ts-nameof/blob/transformationApi/src/main.ts) that does a simple implementation `nameof`.

I think it would be very useful to add the ability to specify "transformation plugins"... maybe something in tsconfig.json. That way people don't have to use the compiler api or a wrapper.

**Comment by p-bakker** [comment]
Hi, this all looks very promising and interesting. 

Is the suggestion that has been made before about allowing to specify the 'transformation plugins'/'modules containing transformations' to use in tsconfig.json anywhere on the roadmap? 

I think it would go a long way in making it easier for the majority of developers to customize their TypeScript > JS transformation if they could just install a transform plugin/module (similar to how you can with typings)

**Comment by mhegazy** [comment]
> Is the suggestion that has been made before about allowing to specify the 'transformation plugins'/'modules containing transformations' to use in tsconfig.json anywhere on the roadmap?

No.

**Comment by p-bakker** [comment]
Hi Mohamed,

Is that due to some reason why it is considered not useful/possible/desired? Or is it just not something that has a high priority?

If technically als logically feasible, would a pullrequest be considered?

**Comment by mhegazy** [comment]
We have no plans to enable a plugin/transform model for the compiler at the time being.  the transforms are available on the compiler API and can be used as such.

**Comment by jwbay** [comment]
> There is an exported function setSourceMapRange that you can use for this

I suppose this means transformers are required to track text ranges and adjust source maps themselves. Would #14441 would help with that (a stab in the dark from the title)? If not, is having user-provided AST transforms automagically adjust source maps planned for the future?

**Comment by mhegazy** [comment]
#14441 is not related to this change. this is infrastructure for code fixes formatting.

As for soruceMaps, sourceMaps are generated by the printer. it needs to track a node to its original source. most of the time you do do not need to do anything, but if you are making a transformation that needs some special mapping you need to set the source range on the new nodes so that the printer knows how to map it.
Take a look at some of the transfomrations we already have today in https://github.com/Microsoft/TypeScript/tree/master/src/compiler/transformers for more details.

**Comment by fabiandev** [comment]
I was playing around with transformations and ran into an issue when replacing a node entirely.

I'm replacing interfaces with symbols, so the following:
```ts
export interface TestInterface { }
```
will become:
```ts
export var TestInterface = Symbol("TestInterface");
```
which works fine.

Now in another file I'm importing the interface as follows:
```ts
import { TestInterface } from './test_interface';
console.log(TestInterface);
```
which becomes:
```ts
console.log(TestInterface);
```
and the import is gone.

I'm using the before transformation with visitors and tried using a transformer for the interface replacement on it's own, before doing other transformations, but it seems the compiler still thinks the imported identifier `TestInterface` is a type, not a variable.

Can anyone point me into the right direction on how to make the compiler aware of this?

(I'm on `typescript@2.3.0-dev.20170313`)

**Comment by mhegazy** [comment]
> and the import is gone.

The decision whether to elide the module import or not is done by the checker before all the transformations have been done. 

I suppose you can replace module transform yourself and not elide imports you care about.

**Comment by fabiandev** [comment]
I see, thanks! I just didn't really get what you meant by:
> replace module transform

**Comment by mhegazy** [comment]
I meant [module transform](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/transformers/module/es2015.ts), We need to expose a way to disable some transforms, it is not possible today.

**Comment by fabiandev** [comment]
Okay, thanks!
I'm now thinking of a way how to work around creating a TypeScript fork to achieve this.

I was first thinking about using `noResolve: true` but that may cause problems later on, so if I'm not mistaken the only option right now is to perform transforms to a source file (using `ts.transform()`), use the printer to create TypeScript code out of it, save it temporarily and to then use the compiler API to actually compile it, so something like:

```ts
const program = ts.createProgram(['./entry.ts'], compilerOptions);
const sourceFiles = program.getSourceFiles().filter(sf => !sf.isDeclarationFile);

const result = ts.transform(sourceFiles, transformerFactories, compilerOptions);
const printer = ts.createPrinter();

for (let transformed of result.transformed) {
  let source = printer.printFile(transformed);
  // save/persist source code temporarily
}

// create a new program with the temporary files and use program.emit
// probably something like: ts.createProgram(['./temp/entry.ts'], compilerOptions)
```

This seems like quite some overhead, and if someone has a better idea I'd be very happy if you could share it!

**Comment by fabiandev** [comment]
I'm experiencing issues with `visitEachChild` for e.g. a variable declaration like this:
```js
let a: () => any;
```

Which results in the following error:
`Error: Debug Failure. False expression: Lexical environment is suspended.`

@rkirov also stated running into issues like that: https://github.com/Microsoft/TypeScript/pull/13940#issuecomment-278821317.

I've reproduced the error on runkit: https://runkit.com/fabiandev/58e571bd9383e400145e1245

@rbuckton @mhegazy Is it a bug or am I missing something?

**Comment by whitecolor** [comment]
A question about API, if one have a TS source text, what API should be used to just format it?

**Comment by rkirov** [comment]
@fabiandev This appears to be a regression from the last time I checked the API. We should open a separate bug for it. A minimal repro is input `function f(a: () => null) {}` with the identity transformer (i.e. just calling visitEachChild with `(n) => n`).

**Comment by fabiandev** [comment]
Thanks @rkirov, I've submitted a separate issue: https://github.com/Microsoft/TypeScript/issues/15192.

**Comment by panesofglass** [comment]
Is it possible to use this API to create declaration files only? In other words, is it possible to create an AST of declarations and generate the d.ts file?

**Comment by mhegazy** [comment]
The declaration file emit is currently not exposed separately on the API. we intend to rewrite the declaration emitter as an emit transformation in the future; once that is done we will expose that on the API as well.

**Comment by pflannery** [comment]
@mhegazy 

> The decision whether to elide the module import or not is done by the checker **before all the transformations have been done**.

When you say "before all transformations have been done", do you mean before custom transformations are executed?

I would like to include new source files in one of my custom transformers. 
My [current transformer](https://github.com/ts-contrib/ts-transformer-aliases/blob/79015d1476656c12ab2331bfb8704f88ca45ee21/src/index.js) replaces module specifiers but I can't find an API to notify the compiler to reprocess new import\referenced files during a transform

**Comment by mhegazy** [comment]
> I would like to include new source files in one of my custom transformers.
My current transformer replaces module specifiers but I can't find an API to notify the compiler to reprocess new import\referenced files during a transform

Transforms, all of them, happen after the checking phase has happened. you can not add new files at this point.

**Comment by EisenbergEffect** [comment]
@mhegazy Did this ever get documented?

**Comment by mhegazy** [comment]
We still have that on our todo list. you can find a lot of samples though in https://github.com/Microsoft/TypeScript/tree/master/src/compiler/transformers

**Comment by SerkanSipahi** [comment]
@EisenbergEffect right :)

We really need documentation like these:
- (user handbook) https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/user-handbook.md
- (plugin handbook) https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md
- (babel types) https://github.com/babel/babel/tree/master/packages/babel-types
- (playground compiler api) https://astexplorer.net (with typescript transformer)

i had written lot of babel plugins for javascript but now im coding for a while with typescript and im missing these documentation for typescript. 

@mhegazy when will you guys start with the documentation?

**Comment by zwwtj2014** [comment]
hello @mhegazy ,

> The declaration file emit is currently not exposed separately on the API. we intend to rewrite the declaration emitter as an emit transformation in the future; once that is done we will expose that on the API as well.

any plan for this?

**Comment by mhegazy** [comment]
Sorry have not had chance to get back to adding the docs yet. https://github.com/Microsoft/TypeScript/issues/14419 tracks doing so. there are a multiple examples for what a transformation looks like in https://github.com/Microsoft/TypeScript/tree/master/src/compiler/transformers

**Comment by SerkanSipahi** [comment]
@mhegazy what i meant was: is Microsoft planning something like this:

We really need documentation like these:
babel:
- (user handbook) https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/user-handbook.md
- (plugin handbook) https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md
- (babel types) https://github.com/babel/babel/tree/master/packages/babel-types
- (playground compiler api) https://astexplorer.net (there is no typescript transformer)

postcss:
- https://github.com/postcss/postcss/blob/master/docs/writing-a-plugin.md
- http://api.postcss.org/

**Comment by mhegazy** [comment]
@SerkanSipahi We are not exposing these transforms on the commandline compiler (like babel does) at the time being, this proposal is tracked by https://github.com/Microsoft/TypeScript/issues/16607.
Currentlly this functionality is only exposed as part of the public API. tools that wrap the compiler, be it ts-loader, or tsickle can expose that functionality.
We do have documentation for the compiler API in https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API, but we still missing documentation for how to use the transform pipeline. that is what #14419 is tracking.

**Comment by cancerberoSgx** [comment]
Here there are three working examples of using transformation APII among other things:  https://typescript-api-playground.glitch.me/. 

BTW you can modify the code and run it again, have intellisense, etc. ...like a TypeScript Compiler API Playground. hope it helps.

**Pull request merged**